### PR TITLE
[CPU][ArmSME] Add (initial) tiling and lowering pipeline for ArmSME

### PR DESCRIPTION
**Pull request by MacDue**
This adds initial support for ArmSME. Currently, this only targets f32 and f64 matmuls (as there's little support for anything else in currently available tools). When `+sme` is set f64 and f32 matmuls will be tiled (at the kernel dispatch level), into sizes that match the SME virtual tiles. The lowerings after that are all standard, until the final LLVM lowerings, where the ArmSME passes are enabled to allow lowering the 2D scalable vector types to ArmSME intrinsics.

Example compiler usage:

```sh
iree-compile \
  --iree-hal-target-backends=llvm-cpu \
  --iree-llvmcpu-target-cpu-features=+sve,+sme \
    matmul.mlir -o $output_file
```

### CODE REVIEW

**Comment by MacDue** [comment]
cc @banach-space, @c-rhodes, @dcaballe 

This is a draft until IREE picks up the required changes in LLVM, I believe that's just:

- ~~https://github.com/llvm/llvm-project/pull/73639~~ :heavy_check_mark: 
- ~~https://github.com/llvm/llvm-project/pull/73931~~ :heavy_check_mark: 
- ~~https://github.com/llvm/llvm-project/pull/73915~~ :heavy_check_mark:

File: compiler/src/iree/compiler/Codegen/LLVMCPU/ConvertToLLVM.cpp
Comment (c-rhodes) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/ConvertToLLVM.cpp]: could the objects attr contain something other than the SME compiler-rt lib? Just wondering if this check should be tighter.

**Review by c-rhodes — COMMENTED** [review]
Thanks for the patch Ben, this is exciting! It's taken a lot of work to get here, mostly in core MLIR, but good to see the integration doesn't require significant changes on the IREE side. Mostly LGTM but I'll take another look once the deps land, cheers

File: compiler/src/iree/compiler/Codegen/LLVMCPU/ConvertToLLVM.cpp
Comment (MacDue) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/ConvertToLLVM.cpp]: I admit this check is a little lazy, it probably should specifically look for the `compiler-rt` object file :)
 I think it's pretty unlikely that something other than `compiler-rt` would be linked though, as I think this is typically used for micro-kernels.

Comment (MacDue) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/ConvertToLLVM.cpp]: I'll fix this up soon :+1:

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (dcaballe) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: doc

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (dcaballe) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:992]: @pzread, another example of an ad-hoc TSS strategy :)

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (dcaballe) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:1053]: Yeah, this after-the-fact tile size adjustments are a pain. We are trying to get rid of those with a new tile size selection infra.

**Review by dcaballe — APPROVED** [review]
Great milestone! LGTM but please wait for Scott or someone else to review the compile_rt dep and the benchmark side of things

File: compiler/src/iree/compiler/Codegen/LLVMCPU/ConvertToLLVM.cpp
Comment (ScottTodd) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/ConvertToLLVM.cpp]: > ArmSME also requires some SME compiler-rt builtins, which need to be provided by setting `IREE_ARM_SME_COMPILER_RT_BUILTINS_STATIC_LIB` to a path that contains them. Though at the moment stubs can also be used instead of the actual builtins.

Can you elaborate on what is included in this static lib? Are the sources for that redistributable? Could we include it as part of our build process, rather than rely on an environment variable? We have similar dependencies for some CPU intrinsics (musl) and GPU SDKs / libraries (CUDA, HIP/ROCm, WebGPU with Tint, etc.).

cc @benvanik , @stellaraccident

Comment (MacDue) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/ConvertToLLVM.cpp]: It's the Arm SME ABI support routines, added to compiler-rt with this commit: https://github.com/llvm/llvm-project/commit/f445be9790f9998e9048fc2a12402f3e213c079f. If a function (or dispatch) is marked as using 'ZA' (the SME accumulator state), the LLVM backend plants calls to ABI support routines to ensure this state is preserved between calls.

Comment (ScottTodd) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/ConvertToLLVM.cpp]: Maybe we could build compiler-rt (when ARM is in `LLVM_TARGETS_TO_BUILD`, see https://github.com/openxla/iree/blob/main/build_tools/cmake/iree_llvm.cmake).

Comment (MacDue) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/ConvertToLLVM.cpp]: I feel it may be a little too early to setup a proper build for `compiler-rt` just yet. We're not actually making use of the functionality it provides yet, so it's simpler to just link in a stub object file. 

Currently, building the actual `compiler-rt` routines with SME requires a pretty recent compiler (clang-16, I believe), otherwise it'll just exclude them from the build. They also depend on either glibc or newlib (depending on if it's a bare-metal build or not), which may complicate things a little more. 

I definitely think we should set this up properly at some point, but I feel it may be simpler to leave this while things are still in flux?

Comment (ScottTodd) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/ConvertToLLVM.cpp]: > We're not actually making use of the functionality it provides yet, so it's simpler to just link in a stub object file.

Could the `return llvmFuncOp.emitError` be downgraded to a warning then? I'm worried that compiling for one CPU architecture will work out of the box but compiling for another will throw an error (and the error would be partway through compilation, not at startup time).

I'd also like to hear from @benvanik or @stellaraccident

Comment (ScottTodd) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/ConvertToLLVM.cpp]: Oh, and if this is left open, can you file an issue to track "setting it up properly"?

Comment (MacDue) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/ConvertToLLVM.cpp]: > Could the return llvmFuncOp.emitError be downgraded to a warning then? I'm worried that compiling for one CPU architecture will work out of the box but compiling for another will throw an error (and the error would be partway through compilation, not at startup time).

It still would throw an error (just during the linking pass).

**Review by benvanik — CHANGES_REQUESTED** [review]
The compiler-rt dep is a hard blocker for this in IREE - we cannot have additional deps like this and what's in it seems to indicate a design issue (not major I think, but at least needs to be solved before landing) if things like abort() and platform-specific (?) register save/restore are being inserted during code generation. IREE executables are portable across all platforms and are never allowed to abort() or make any kind of syscalls or globally impacting operations.

Ideally upstream shouldn't be inserting aborts in user programs, or should be inserting them in a way we can remove them here by turning them into non-zero returns from the dispatch functions (gracefully failing instead of bringing down the entire user process). I suspect this isn't too hard to fix but may require a pass/pattern to do that substitution if upstream is really inserting aborts (find your aborts and replace with return 1 may be enough). That pass could live in IREE if needed but we'd want to have a stable thing to match on emitted by the upstream code.

How platform-specific is the register save/restore code? If it's always the same on all operating systems (and the platform-specific comments in the compiler-rt impl are just about the abort) then it may just mean moving that out of compiler-rt and into LLVM IR you emit. We do have a mechanism for builtins (https://github.com/openxla/iree/blob/main/runtime/src/iree/builtins/device/README.md) that we link into codegen binaries where the save/restore could be moved so long as they are portable, but IMO not requiring any deps or users to do such things for what amounts to a dozen assembly instructions would be ideal and help with hermetic deployments. This again may be a pass we carry in IREE that just adds those two functions in LLVM IR with inline asm blocks only when one or more dispatches uses SME.

Happy to chat more about the exact details, but the TLDR is no abort() (must return non-zero if failures are unrecoverable) and no compiler-rt dep for save/restore. There are ways around both and you may have some approaches you've discussed already. It's a blocker here but I don't think a large lift to have a pass that does the right thing, and we're happy to carry it if needed. Happy to discuss more on the iree-discuss list, discord, or the next community meeting!

**Comment by c-rhodes** [comment]
> How platform-specific is the register save/restore code? If it's always the same on all operating systems (and the platform-specific comments in the compiler-rt impl are just about the abort) then it may just mean moving that out of compiler-rt and into LLVM IR you emit. We do have a mechanism for builtins (https://github.com/openxla/iree/blob/main/runtime/src/iree/builtins/device/README.md) that we link into codegen binaries where the save/restore could be moved so long as they are portable, but IMO not requiring any deps or users to do such things for what amounts to a dozen assembly instructions would be ideal and help with hermetic deployments. This again may be a pass we carry in IREE that just adds those two functions in LLVM IR with inline asm blocks only when one or more dispatches uses SME.

Thanks for the detailed reply Ben. I think for now at least stubs would be sufficient for these routines, we can use the builtins mechanism you shared for this.

For a bit more context, these routines are to support the ZA (matrix register state defined in SME) lazy save scheme [1], which intends to reduce the number of saves and restores of ZA state between calls. However, this is only relevant when  either caller/callee may have live ZA data or in a multi-threaded context where there is contention for this state.

For the former, the likelihood of this in IREE where we're using the ZA in dispatch functions seems slim. I think we've only seen dispatch functions call other functions in one of our pipelines, and in this case it was libc routines which wouldn't be a concern here (since they wont be using ZA).

The latter is of concern, but not at the moment for us as we're concentrating on single-threaded currently, but this will become more important for us in the future.

[1] https://github.com/ARM-software/abi-aa/blob/main/aapcs64/aapcs64.rst#the-za-lazy-saving-scheme

**Comment by MacDue** [comment]
I've updated the patch to add Arm SME ABI stubs to `libdevice` (which should be sufficient for now).

I think that's a much nicer setup, thanks for the advice! :+1: 

Just a small thing to note:
> or in a multi-threaded context where there is contention for this state.

I believe between OS threads it's the kernel's job to save/restore the ZA state (see https://docs.kernel.org/arch/arm64/sme.html). The routines added here are just for saving the state between function calls. Since SME-enabled dispatch regions are self-contained (and don't call out to other SME-enabled functions), we know these ABI functions won't end up being called by the code IREE generates. The backend still plants these calls, since it does not know how the functions will be called, but with how things are currently set up we won't ever need to save our caller's ZA state (for example).

**Comment by MacDue** [comment]
I think it's also possible to implement these routines via inline assembly in a C file, but as I don't think they're actually needed, stubs seem preferable. 

w.r.t. the `abort()`s, I think those are only reachable via an incorrect program. We can patch the assembly just to return in that case anyway, but it's not possible to return a value. The ABI functions are all defined to return `void` in the standard, and the functions the calls are added to don't really know they exist (I think they're inserted quite late in the LLVM backend). If you wanted to know they failed I think they'd need to call out to some runtime function to report the error. So instead of an abort, maybe call something like `iree_report_execution_error_in_current_thread()`? 

But as previously mentioned, I think it's safe for these functions to just be stubs for the code we currently generate.

File: runtime/src/iree/builtins/device/BUILD.bazel
Comment (c-rhodes) [inline runtime/src/iree/builtins/device/BUILD.bazel]: `libdevice_aarch64_sme` for consistency?

File: runtime/src/iree/builtins/device/arch/arm_64/sme_abi.c
Comment (c-rhodes) [inline runtime/src/iree/builtins/device/arch/arm_64/sme_abi.c]: Given we don't expect SME-enabled dispatch functions with live ZA state to call other such functions, I think we can remove the TODO's, as these stubs are sufficient?

File: runtime/src/iree/builtins/device/BUILD.bazel
Comment (MacDue) [reply runtime/src/iree/builtins/device/BUILD.bazel]: Whoops, that was a typo! I was blind to the double 'device' there :sweat_smile: 
I'll fix that up 👍

File: runtime/src/iree/builtins/device/arch/arm_64/sme_abi.c
Comment (MacDue) [reply runtime/src/iree/builtins/device/arch/arm_64/sme_abi.c]: I think it's still a TODO, even if we don't need these functions (yet), that might not always be the case.

**Review by c-rhodes — APPROVED** [review]
LGTM cheers

File: runtime/src/iree/builtins/device/arch/arm_64/sme_abi.c
Comment (benvanik) [reply runtime/src/iree/builtins/device/arch/arm_64/sme_abi.c]: with TODOs we try to make them actionable (in case the originating author disappears) - a comment of "// No-op today as not needed." would be informative/durable and let people know this file is good as-is, while the TODO without context makes it look like it's not done yet.

**Review by benvanik — APPROVED** [review]
Nice! Thanks for the changes. We can evaluate in the future if we move these into the compiler to emit instead of needing a unique bitcode file but this at least puts things in the control of the compiler and will be easy to fix up in the future. One nit on the TODOs and otherwise LGTM!

File: tests/e2e/matmul/BUILD.bazel
Comment (hanhanW) [inline tests/e2e/matmul/BUILD.bazel:67]: I think it is okay to add a new entry for the test suite. But can we remove the original one? E.g., can we remove line 50 in this file?

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:1053]: I thought the "infra" is more like providing you a way to configure tile sizes. The adjustment is quite independent to the infra. The size adjustment is more about how you compute tile sizes. You can always get the tile sizes you set without adjustment, just don't call those functions?

Anyway, I agree that it is a pain; I've been working progress on removing historical heuristics. I just wanna flag that you will run into the same situation even we have a new tile size selection infra. You will either call the existing methods or re-implement same methods in a different way. The new tile size selection infra does not address the adjustment issue to me. It is an infra, but not a "fix".

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: Do we need to check `linalg:FillOp`? It is not common to have a single fill op in dispatch. I thought this function is mostly used for matmuls?

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: Can we remove the `.empty()` check? I think this is an invalid linalg op if it happens.

**Review by hanhanW — APPROVED** [review]
Awesome work!

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (pzread) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:1053]: I think one point we can improve is not having these target-dependent decisions in the generic utilities like `getMatmulVectorSizes` here (and [this](https://github.com/openxla/iree/blob/2529fb37c76c89c0306a4bb45f8bdcdf8baeaf16/compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp#L1057-L1062)). This can be one reason confusing as we are trying to have a generic tile size selection workflow for each op, but target-specific logic/parameters/heuristics are segregated in each step (distribution, cache, vector tile selection steps)

I'm trying something like dispatching the tile size selection workflow based on the target architecture, then each workflow only needs to take care about its target-specific parameters/logic/heuristics. Then we can factor out common code into target-independent utilities so each workflow can be implemented concisely. Nothing concretes yet but just an idea.

**Review by ScottTodd — APPROVED** [review]

**Comment by dcaballe** [comment]
Awesome! Thank you all!

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (MacDue) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: Nope, good point, this only applies to `LinalgContractionOps`, so won't ever see a fill :+1:

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (MacDue) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: I've changed this to a debug assert :+1:

File: tests/e2e/matmul/BUILD.bazel
Comment (MacDue) [reply tests/e2e/matmul/BUILD.bazel:67]: The tests added here are just for `SME`, line 50 adds tests for `SVE`. I put the `SME` tests in their own section as (for now), SME is a bit more of a special case :)

**Comment by MacDue** [comment]
Thanks everyone for the reviews! :)

File: tests/e2e/matmul/BUILD.bazel
Comment (hanhanW) [reply tests/e2e/matmul/BUILD.bazel:67]: oops, I misread it. Sorry about that and good catch!

**Pull request merged**
